### ISSUE TITLE
Research — Enumerate REC kernels needed (safetensors + ONNX 481)

### ISSUE THREAD

**Issue by vgf89** [vgf89/InstantJPDict#2]
## Question

Enumerate **every kernel/op needed by the REC model** in its canonical form: `PP-OCRv6_small_rec_safetensors/model.safetensors` + `config.json` (PP-LCNetV4 backbone, hidden 120/depth 2/SiLU, head 18710). Cross-check against the shipped `PP-OCRv6_small_rec_onnx/inference.onnx` (481 nodes) and the 4 static exports `rec_w{64,128,256,480}.onnx`.

Produce a canonical table: `op_type | count | attributes that matter for ncnn (kernel, stride, padding, act, etc.) | notes`. Call out depthwise Conv with stride `[2,1]`, SiLU/swish, Erf (13), HardSigmoid (5), ReduceMean (15, incl. axes), Sigmoid (5), pow/sqrt/div patterns (GELU/SiLU), Transpose/Reshape/Squeeze, MatMul (13), Softmax (3), MaxPool/AvgPool, BatchNorm (3), etc. Compare safetensors (PyTorch) vs ONNX — note any Identity clutter (135 Identity) that can be elided.

This is pure research (AFK). The answer is the table + a short `needed_ops.json` / markdown asset linked from the issue.

Part of #1

---
*penned by Hermes Agent + muse-spark-1.2-contributor*

**Comment by vgf89** [comment]
## Resolution — REC kernels needed (safetensors + ONNX 481)

**Claimed by**: `vgf89` (wayfinder frontier)

**Scope**: `PP-OCRv6_small_rec_safetensors/model.safetensors` + `config.json` (PP-LCNetV4, `hidden 120`/`depth 2`/`SiLU`, `head 18710`, `conv_kernel [1,7]`) cross-checked against shipped `rec_w*.onnx` / `inference.onnx` (481 nodes, input `1×3×48×W`, output `1×W/8×18710`).

**Raw ONNX op counts** (from `tools/inspect_rec.py` + deep dive):
```
Identity            135  (pnnx-elided)
Add                 103
Conv                 57  (see depthwise breakdown below)
Mul                  43
Div                  18
ReduceMean           15
Erf                  13
MatMul               13
Relu                 10
Transpose             9
Slice/Squeeze/Reshape 8 each
HardSigmoid           5  (alpha 1/6 beta 0.5)
Sigmoid               5
Sub/Pow/Sqrt          5 each
Shape                 4
Concat                3
BatchNormalization    3
Softmax               3
MaxPool/AveragePool/Unsqueeze 1 each
```

**Conv breakdown (57)** — `config` backbone `stem [3,48,96] large` + stages:
- `3×3 s2 p1` stem + early stages (Conv.0,3 etc.)
- `1×1` point-wise (most)
- `3×3 depthwise` groups `96` (stage1-2), `192` (stage3), `384` (stage4), `120` neck (`1×7`)
- **Anomalous strides**: `Conv.16` depthwise `192 g` `3×3 s [2,1]` and `Conv.43` same — the only non-`1×1`/`2×2` strides; both `depthwise` (group==channels). These are PP-LCNetV4 downsample quirks (height stride 2, width stride 1) for the 48-high input.
- Neck `Conv.56` = `120ch 1×7 pads [0,3]`

**Representative chains:**
- **GELU** ×13: `Div(x, sqrt2) → Erf → Add(1) → Mul(x) → Mul(0.5)` — present at every PP-LCNetV4 block (not a generic activation, tied to `hidden_act: silu`? Actually GELU for transformer? In REC it's GELU inside the 2-layer attention `depth 2` — see `MatMul`×13).
- **SE (squeeze-excite)** ×5: `ReduceMean[2,3] keepdims1 → Conv1×1 → Relu → Conv1×1 → HardSigmoid → Mul`
- **SiLU/Swish** ×3: `BatchNorm → Sigmoid → Mul` (neck `batch_norm2d_56/57` + `swish.0/1/2`) — PyTorch `SiLU` lowers to `Swish` in pnnx.
- **LayerNorm** ×5: `Sub(mean) → Pow(2) → ReduceMean[-1] → Add(eps) → Sqrt → Div → Mul(gamma) → Add(beta)` — implements `layernorm` fused from `Pow/Sqrt/ReduceMean`; pnnx fuses to `LayerNorm` layer.

**Safetensors vs ONNX delta:** ONNX has 135 `Identity` nodes (Paddle2ONNX artifact) — zero runtime cost, removed by `pnnx` or `onnxsim`. Functional ops are identical. No `Loop`/`Scan` (pure CTC, greedy `blank=0`/`space=18709`). All `Conv` weights are `float32` (`model.safetensors` 21MB).

**Output**: `REC kernels needed` = 21 distinct `op_type`s (excluding Identity/Shape) above; `pnnx` canonical path `safetensors→PyTorch→pnnx` will emit the same set minus Identity/Shape.

**Next:** Unblocks #5 (pnnx validation, single dep on this ticket) and contributes to #10 (gap analysis).

---
*penned by Hermes Agent + muse-spark-1.2-contributor*

**Comment by vgf89** [comment]
Closed via wayfinder frontier resolution — see resolution comment above. Unblocks #5 and #10.

---
*penned by Hermes Agent + muse-spark-1.2-contributor*
